helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -152,69 +179,147 @@ private double getAssignmentNormalizedScore(AssignableNode node, AssignableRepli
-      String resourceName1 = replica1.getResourceName();
-      String resourceName2 = replica2.getResourceName();
-      if (bestPossibleAssignment.containsKey(resourceName1) == bestPossibleAssignment
-          .containsKey(resourceName2)) {
-        if (baselineAssignment.containsKey(resourceName1) == baselineAssignment
-            .containsKey(resourceName2)) {
-          // If both assignment states have/not have the resource assignment the same,
-          // compare for additional dimensions.
-          int statePriority1 = replica1.getStatePriority();
-          int statePriority2 = replica2.getStatePriority();
-          if (statePriority1 == statePriority2) {
-            // If state priorities are the same, try to randomize the replicas order. Otherwise,
-            // the same replicas might always be moved in each rebalancing. This is because their
-            // placement calculating will always happen at the critical moment while the cluster is
-            // almost close to the expected utilization.
-            //
-            // Note that to ensure the algorithm is deterministic with the same inputs, do not use
-            // Random functions here. Use hashcode based on the cluster topology information to get
-            // a controlled randomized order is good enough.
-            Integer replicaHash1 = replicaHashCodeMap.get(replica1.toString());
-            Integer replicaHash2 = replicaHashCodeMap.get(replica2.toString());
-            if (!replicaHash1.equals(replicaHash2)) {
-              return replicaHash1.compareTo(replicaHash2);
-            } else {
-              // In case of hash collision, return order according to the name.
-              return replica1.toString().compareTo(replica2.toString());
-            }
-          } else {
-            // Note we shall prioritize the replica with a higher state priority,
-            // the smaller priority number means higher priority.
-            return statePriority1 - statePriority2;
-          }
-        } else {
-          // If the baseline assignment contains the assignment, prioritize the replica.
-          return baselineAssignment.containsKey(resourceName1) ? -1 : 1;
+  private Pair<AssignableReplica, List<AssignableReplica>> getNextAssignableReplica(
+      List<AssignableReplica> allReplica,
+      Map<String, MutablePair<Integer, Integer>> overallClusterUtilMap, ClusterModel clusterModel,
+      Map<String, Integer> _replicaHashCodeMap) {
+    List<AssignableReplica> replicaForNextRound = new ArrayList<>();
+    float highestScore = -1;
+    AssignableReplica nextAssinableReplica = null;
+    Map<String, Float> weightMap = new HashMap<>();
+    // compute weight for each capacity dimension
+    for (Map.Entry<String, MutablePair<Integer, Integer>> resourceUsage : overallClusterUtilMap
+        .entrySet()) {
+      float freePercentage =
+          (float) (resourceUsage.getValue().right - resourceUsage.getValue().left) / resourceUsage
```

Comment:
For protecting any possible future errors (for example, we allow overprovisioning in the future), please bound the subtraction result it should be 0 at least, I guess. Or if the later calculation can handle a negative percentage number, then it would be fine here.